File: scripts/platforms/steam.js
// Provider Steam.
//
// Contrat commun à toutes les plateformes (voir scripts/library.js) :
//   id, name  -> identité de la plateforme
//   scan()    -> { games: [...] }
// Chaque jeu retourné DOIT porter : platform, id, name, launch, art.
//
// Cette logique sera reportée côté Rust une fois le scaffold Tauri en place.

const ID = 'steam';
const NAME = 'Steam';

const fs = require('fs');
const path = require('path');
const os = require('os');

/**
 * Parse minimal du format VDF de Valve (KeyValues).
 * Le format est un arbre clé/valeur : "cle" "valeur" ou "cle" { ... }
 * On ne gère que ce dont on a besoin — pas de dépendance externe.
 */
function parseVDF(text) {
  const root = {};
  const stack = [root];
  // Capture soit une paire "cle" "valeur", soit "cle" seule (ouverture de bloc), soit { }
  const re = /"((?:[^"\\]|\\.)*)"\s*(?:"((?:[^"\\]|\\.)*)")?|(\{)|(\})/g;
  let m;
  let pendingKey = null;

  while ((m = re.exec(text)) !== null) {
    const [, key, value, open, close] = m;

    if (open) {
      const node = {};
      if (pendingKey !== null) {
        stack[stack.length - 1][pendingKey] = node;
        pendingKey = null;
      }
      stack.push(node);
    } else if (close) {
      if (stack.length > 1) stack.pop();
    } else if (key !== undefined) {
      if (value !== undefined) {
        stack[stack.length - 1][key] = value;
        pendingKey = null;
      } else {
        pendingKey = key;
      }
    }
  }
  return root;
}

/**
 * Tout ce que Steam installe mais qui n'est pas un jeu : runtimes, redistribuables, outils.
 * Ils ont un appmanifest exactement comme un jeu — seul un filtre explicite les distingue.
 */
const NON_GAME_APPIDS = new Set([
  '228980',  // Steamworks Common Redistributables
  '1070560', // Steam Linux Runtime
  '1391110', // Steam Linux Runtime - Soldier
  '1628350', // Steam Linux Runtime 3.0 (Sniper)
  '1493710', // Proton Experimental
  '2180100', // Proton Hotfix
]);
const NON_GAME_NAME = /redistributable|steam linux runtime|^proton\b|^steamworks/i;

function isGame(g) {
  return !NON_GAME_APPIDS.has(String(g.appid)) && !NON_GAME_NAME.test(g.name);
}

/** Trouve le dossier d'installation de Steam. */
function findSteamRoot() {
  const candidates = [
    path.join(process.env['ProgramFiles(x86)'] || 'C:\\Program Files (x86)', 'Steam'),
    path.join(process.env['ProgramFiles'] || 'C:\\Program Files', 'Steam'),
    'C:\\Steam',
    path.join(os.homedir(), '.steam', 'steam'), // Linux, pour plus tard
  ];
  return candidates.find(c => fs.existsSync(path.join(c, 'steamapps', 'libraryfolders.vdf'))) || null;
}

/** Liste tous les dossiers de bibliothèque (Steam peut en avoir sur plusieurs disques). */
function findLibraryFolders(steamRoot) {
  const vdfPath = path.join(steamRoot, 'steamapps', 'libraryfolders.vdf');
  const parsed = parseVDF(fs.readFileSync(vdfPath, 'utf8'));
  const folders = parsed.libraryfolders || parsed.LibraryFolders || {};

  const paths = [];
  for (const [key, val] of Object.entries(folders)) {
    if (!/^\d+$/.test(key)) continue; // les entrées sont indexées "0", "1", "2"...
    const p = typeof val === 'string' ? val : val.path;
    if (p) paths.push(p.replace(/\\\\/g, '\\'));
  }
  // Le dossier racine n'est pas toujours listé
  if (!paths.some(p => path.resolve(p) === path.resolve(steamRoot))) paths.unshift(steamRoot);
  return paths;
}

/** Lit les appmanifest_*.acf d'une bibliothèque et en sort les jeux installés. */
function readGames(libraryPath) {
  const appsDir = path.join(libraryPath, 'steamapps');
  if (!fs.existsSync(appsDir)) return [];

  return fs.readdirSync(appsDir)
    .filter(f => /^appmanifest_\d+\.acf$/.test(f))
    .map(f => {
      try {
        const state = parseVDF(fs.readFileSync(path.join(appsDir, f), 'utf8')).AppState;
        if (!state || !state.appid || !state.name) return null;
        return {
          platform: ID,                 // d'où vient le jeu — jamais implicite
          id: String(state.appid),      // identifiant DANS sa plateforme
          name: state.name,
          launch: `steam://rungameid/${state.appid}`,
          installDir: state.installdir,
          sizeBytes: Number(state.SizeOnDisk || 0),
          lastPlayed: Number(state.LastPlayed || 0), // epoch — notre tri par récence
          library: libraryPath,
        };
      } catch {
        return null; // un manifeste corrompu ne doit pas faire tomber tout le scan
      }
    })
    .filter(Boolean);
}

/** Collecte récursivement les fichiers sous `dir`, jusqu'à `depth` niveaux. */
function collectFiles(dir, depth, out) {
  if (depth === 0) return;
  let entries;
  try { entries = fs.readdirSync(dir, { withFileTypes: true }); } catch { return; }
  for (const e of entries) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) collectFiles(p, depth - 1, out);
    else out.push(p);
  }
}

// `library_hero_blur.jpg` contient `library_hero` : sans exclusion explicite,
// on servirait la version floue à la place du vrai visuel.
const KIND = {
  portrait: f => f.includes('library_600x900'),
  hero: f => f.includes('library_hero') && !f.includes('library_hero_blur'),
  logo: f => f.includes('logo'),
};

/**
 * Steam a empilé trois structures de cache au fil des versions :
 *   librarycache/<appid>_library_hero.jpg          (à plat, ancien)
 *   librarycache/<appid>/library_hero.jpg          (2 niveaux)
 *   librarycache/<appid>/<hash>/library_hero.jpg   (3 niveaux, récent)
 * Les trois coexistent sur une même machine — d'où la recherche récursive.
 */
function findArtwork(steamRoot, appid) {
  const cache = path.join(steamRoot, 'appcache', 'librarycache');
  const found = {};
  if (!fs.existsSync(cache)) return found;

  const candidates = [];

  // Structures 2 et 3 niveaux : tout ce qui vit sous librarycache/<appid>/
  const appDir = path.join(cache, String(appid));
  if (fs.existsSync(appDir)) collectFiles(appDir, 3, candidates);

  // Structure à plat : librarycache/<appid>_library_hero.jpg
  try {
    const prefix = `${appid}_`;
    for (const e of fs.readdirSync(cache, { withFileTypes: true })) {
      if (e.isFile() && e.name.startsWith(prefix)) candidates.push(path.join(cache, e.name));
    }
  } catch { /* cache illisible : on repart avec ce qu'on a */ }

  for (const [kind, matches] of Object.entries(KIND)) {
    const hit = candidates.find(p => {
      const f = path.basename(p).toLowerCase();
      return /\.(jpg|png)$/.test(f) && matches(f);
    });
    if (hit) found[kind] = hit;
  }
  return found;
}

/** Détecte si la plateforme est présente sur la machine, sans lever d'erreur. */
function detect() {
  return findSteamRoot() !== null;
}

function scan() {
  const steamRoot = findSteamRoot();
  if (!steamRoot) throw new Error('Steam introuvable sur cette machine.');

  const libraries = findLibraryFolders(steamRoot);
  const games = libraries.flatMap(readGames).filter(isGame);
  for (const g of games) g.art = findArtwork(steamRoot, g.id);

  return { root: steamRoot, libraries, games };
}

module.exports = { id: ID, name: NAME, detect, scan, parseVDF, isGame };

File: src-tauri/src/platforms/steam.rs
//! Provider Steam. Port Rust de scripts/platforms/steam.js — même logique,
//! même filtre anti-faux-jeux, même recherche de jaquettes dans le cache local.

use crate::library::{Art, Game};
use crate::vdf;
use std::fs;
use std::path::{Path, PathBuf};

pub const ID: &str = "steam";
pub const NAME: &str = "Steam";

/// Ce que Steam installe mais qui n'est pas un jeu : runtimes, redistribuables.
/// Ils ont un appmanifest exactement comme un jeu — seul un filtre explicite les distingue.
const NON_GAME_APPIDS: &[&str] = &[
    "228980",  // Steamworks Common Redistributables
    "1070560", // Steam Linux Runtime
    "1391110", // Steam Linux Runtime - Soldier
    "1628350", // Steam Linux Runtime 3.0 (Sniper)
    "1493710", // Proton Experimental
    "2180100", // Proton Hotfix
];

fn is_non_game_name(name: &str) -> bool {
    let n = name.to_lowercase();
    n.contains("redistributable")
        || n.contains("steam linux runtime")
        || n.starts_with("proton")
        || n.starts_with("steamworks")
}

fn find_steam_root() -> Option<PathBuf> {
    let pf86 = std::env::var("ProgramFiles(x86)").unwrap_or_else(|_| r"C:\Program Files (x86)".into());
    let pf = std::env::var("ProgramFiles").unwrap_or_else(|_| r"C:\Program Files".into());
    let candidates = [
        PathBuf::from(format!("{pf86}\\Steam")),
        PathBuf::from(format!("{pf}\\Steam")),
        PathBuf::from(r"C:\Steam"),
    ];
    candidates
        .into_iter()
        .find(|c| c.join("steamapps").join("libraryfolders.vdf").exists())
}

pub fn detect() -> bool {
    find_steam_root().is_some()
}

/// Steam peut avoir des bibliothèques sur plusieurs disques.
fn find_library_folders(steam_root: &Path) -> Vec<PathBuf> {
    let vdf_path = steam_root.join("steamapps").join("libraryfolders.vdf");
    let text = match fs::read_to_string(&vdf_path) {
        Ok(t) => t,
        Err(_) => return vec![steam_root.to_path_buf()],
    };
    let parsed = vdf::parse(&text);

    let mut paths = Vec::new();
    if let Some(folders) = parsed.get("libraryfolders").and_then(|v| v.as_map()) {
        for (key, val) in folders {
            if !key.chars().all(|c| c.is_ascii_digit()) {
                continue; // les entrées sont indexées "0", "1", "2"...
            }
            let p = match val {
                vdf::Value::Str(s) => Some(s.clone()),
                vdf::Value::Map(m) => m.get("path").and_then(|v| v.as_str()).map(str::to_string),
            };
            if let Some(p) = p {
                paths.push(PathBuf::from(p));
            }
        }
    }
    if !paths.iter().any(|p| p == steam_root) {
        paths.insert(0, steam_root.to_path_buf()); // le dossier racine n'est pas toujours listé
    }
    paths
}

/// Collecte récursivement les fichiers sous `dir`, jusqu'à `depth` niveaux.
fn collect_files(dir: &Path, depth: usize, out: &mut Vec<PathBuf>) {
    if depth == 0 {
        return;
    }
    if let Ok(entries) = fs::read_dir(dir) {
        for e in entries.filter_map(|e| e.ok()) {
            let p = e.path();
            if p.is_dir() {
                collect_files(&p, depth - 1, out);
            } else {
                out.push(p);
            }
        }
    }
}

/// `library_hero_blur.jpg` contient `library_hero` : sans exclusion explicite,
/// on servirait la version floue à la place du vrai visuel.
fn is_kind(fname: &str, kind: &str) -> bool {
    match kind {
        "portrait" => fname.contains("library_600x900"),
        "hero" => fname.contains("library_hero") && !fname.contains("library_hero_blur"),
        "logo" => fname.contains("logo"),
        _ => false,
    }
}

/// Steam a empilé trois structures de cache au fil des versions :
///   librarycache/<appid>_library_hero.jpg          (à plat, ancien)
///   librarycache/<appid>/library_hero.jpg          (2 niveaux)
///   librarycache/<appid>/<hash>/library_hero.jpg   (3 niveaux, récent)
/// Les trois coexistent sur une même machine — d'où la recherche récursive.
fn find_artwork(steam_root: &Path, appid: &str) -> Art {
    let cache = steam_root.join("appcache").join("librarycache");
    if !cache.exists() {
        return Art::default();
    }

    let mut candidates: Vec<PathBuf> = Vec::new();

    // Structures 2 et 3 niveaux : tout ce qui vit sous librarycache/<appid>/
    let app_dir = cache.join(appid);
    if app_dir.is_dir() {
        collect_files(&app_dir, 3, &mut candidates);
    }

    // Structure à plat : librarycache/<appid>_library_hero.jpg
    if let Ok(entries) = fs::read_dir(&cache) {
        let prefix = format!("{appid}_");
        for e in entries.filter_map(|e| e.ok()) {
            if e.file_name().to_string_lossy().starts_with(&prefix) {
                let p = e.path();
                if p.is_file() {
                    candidates.push(p);
                }
            }
        }
    }

    let pick = |kind: &str| -> Option<String> {
        candidates
            .iter()
            .find(|p| {
                let f = p
                    .file_name()
                    .map(|n| n.to_string_lossy().to_lowercase())
                    .unwrap_or_default();
                (f.ends_with(".jpg") || f.ends_with(".png")) && is_kind(&f, kind)
            })
            .map(|p| p.to_string_lossy().to_string())
    };

    Art {
        portrait: pick("portrait"),
        hero: pick("hero"),
        logo: pick("logo"),
    }
}

fn read_games(steam_root: &Path, library_path: &Path) -> Vec<Game> {
    let apps_dir = library_path.join("steamapps");
    let entries = match fs::read_dir(&apps_dir) {
        Ok(e) => e,
        Err(_) => return vec![],
    };

    entries
        .filter_map(|e| e.ok())
        .filter_map(|entry| {
            let fname = entry.file_name().to_string_lossy().to_string();
            if !(fname.starts_with("appmanifest_") && fname.ends_with(".acf")) {
                return None;
            }
            // Un manifeste corrompu ne doit pas faire tomber tout le scan.
            let text = fs::read_to_string(entry.path()).ok()?;
            let parsed = vdf::parse(&text);
            let state = parsed.get("AppState")?.as_map()?;
            let appid = state.get("appid")?.as_str()?.to_string();
            let name = state.get("name")?.as_str()?.to_string();
            let size_bytes: u64 = state
                .get("SizeOnDisk")
                .and_then(|v| v.as_str())
                .and_then(|s| s.parse().ok())
                .unwrap_or(0);
            let last_played: i64 = state
                .get("LastPlayed")
                .and_then(|v| v.as_str())
                .and_then(|s| s.parse().ok())
                .unwrap_or(0);

            Some(Game {
                platform: ID.to_string(),
                id: appid.clone(),
                name,
                launch: format!("steam://rungameid/{appid}"),
                size_bytes,
                last_played,
                art: find_artwork(steam_root, &appid),
            })
        })
        .filter(|g| !NON_GAME_APPIDS.contains(&g.id.as_str()) && !is_non_game_name(&g.name))
        .collect()
}

pub fn scan() -> Result<Vec<Game>, String> {
    let steam_root = find_steam_root().ok_or_else(|| "Steam introuvable sur cette machine.".to_string())?;
    let libraries = find_library_folders(&steam_root);
    let mut games: Vec<Game> = libraries.iter().flat_map(|lib| read_games(&steam_root, lib)).collect();
    games.sort_by(|a, b| b.last_played.cmp(&a.last_played));
    Ok(games)
}
